### PR TITLE
add charset to content-type headers in multipart body

### PR DESCRIPTION
**Pull request by dkarwacki**
Before submitting pull request:
- [ ] Check if the project compiles by running `sbt compile`
- [ ] Verify docs compilation by running `sbt compileDocs`
- [ ] Check if tests pass by running `sbt test`
- [ ] Format code by running `sbt scalafmt`

### CODE REVIEW

File: core/src/test/scalajs/sttp/client3/testing/AbstractFetchHttpTest.scala [hunk 1]
```diff
@@ -27,6 +27,8 @@ abstract class AbstractFetchHttpTest[F[_], +P] extends HttpTest[F] {
 
   override protected def supportsHostHeaderOverride = false
 
+  override protected def supportsCharsetControl = false
```
Comment (adamw) [inline core/src/test/scalajs/sttp/client3/testing/AbstractFetchHttpTest.scala:30]: This should probably say sth about multiparts, e.g. `supportsCharsetsInMultipartParts`

File: core/src/main/scala/sttp/client3/internal/package.scala [hunk 2]
```diff
@@ -30,6 +34,18 @@ package object internal {
     transfer()
   }
 
+  private[client3] def withCharset(headers: Map[String, String], encoding: String): Map[String, String] = {
+    val contentTypeOpt = for {
+      charset <- Try(Charset.forName(encoding)).toOption
```
Comment (adamw) [inline core/src/main/scala/sttp/client3/internal/package.scala:39]: do we need this validation? if the user wants to specify an invalid charset, I'd let them do so. Or throw an exception, but that might be too restrictive

File: core/src/main/scalajvm/sttp/client3/HttpURLConnectionBackend.scala [hunk 3]
```diff
@@ -150,7 +150,13 @@ class HttpURLConnectionBackend private (
     // inspired by: https://github.com/scalaj/scalaj-http/blob/master/src/main/scala/scalaj/http/Http.scala#L542
     val partsWithHeaders = mp.parts.map { p =>
       val contentDisposition = s"${HeaderNames.ContentDisposition}: ${p.contentDispositionHeaderValue}"
-      val otherHeaders = p.headers.map(h => s"${h.name}: ${h.value}")
+      val otherHeaders = p.body match {
+        case StringBody(_, encoding, _) =>
+          withCharset(p.headers.map(h => h.name -> h.value).toMap, encoding).map { case (name, value) =>
```
Comment (adamw) [inline core/src/main/scalajvm/sttp/client3/HttpURLConnectionBackend.scala:155]: ah, in `Part` you already have a `Seq[Header]` so you can use that to create a `Headers` instance

File: core/src/main/scala/sttp/client3/internal/package.scala [hunk 4]
```diff
@@ -30,6 +34,18 @@ package object internal {
...
+      contentType <- headers.get(HeaderNames.ContentType) if !contentType.contains("charset")
```
Comment (adamw) [inline core/src/main/scala/sttp/client3/internal/package.scala:40]: headers need to be compare using case-insensitive checks. There's a `Header.is` helper method, as well as a helper `Headers` wrapper, which even has a `.contentType` method on it. Maybe we can use it here.

File: core/src/main/scala/sttp/client3/internal/package.scala [hunk 5]
```diff
@@ -30,6 +34,18 @@ package object internal {
     transfer()
   }
 
+  private[client3] def withCharset(headers: Map[String, String], encoding: String): Map[String, String] = {
```
Comment (adamw) [inline core/src/main/scala/sttp/client3/internal/package.scala:37]: maybe we could add an `addCharsetToContentType(charset: String): Part` method on `Part`, that could nicely encapsulate that logic

File: core/src/main/scala/sttp/client3/internal/package.scala (hunk 2 above)
Comment (dkarwacki) [reply core/src/main/scala/sttp/client3/internal/package.scala:39]: but the user specifies the encoding, and we are changing it to the charset behind the scenes, so it's maybe better to keep content-type untouched if the encoding is incorrect?

Comment (adamw) [reply core/src/main/scala/sttp/client3/internal/package.scala:39]: I think silent errors might be very confusing

Comment (dkarwacki) [reply core/src/main/scala/sttp/client3/internal/package.scala:39]: ok, so i will get rid of it

File: finagle-backend/src/main/scala/sttp/client3/finagle/FinagleBackend.scala [hunk 6]
```diff
@@ -128,11 +128,7 @@ class FinagleBackend(client: Option[Client] = None) extends SttpBackend[TFuture,
       case MultipartBody(_)       => throw new IllegalArgumentException("Nested multipart bodies are not supported")
     }
 
-    part.fileName match {
-      case Some(_) => FileElement(part.name, ByteArray(content.getBytes: _*), part.contentType, part.fileName)
-      case None    => SimpleElement(part.name, content)
-    }
-
+    FileElement(part.name, ByteArray(content.getBytes: _*), part.contentType, part.fileName)
```
Comment (adamw) [inline finagle-backend/src/main/scala/sttp/client3/finagle/FinagleBackend.scala:131]: I don' think we always want to return a file element?

File: core/src/test/scala/sttp/client3/testing/HttpTest.scala [hunk 7]
```diff
@@ -409,6 +410,28 @@ trait HttpTest[F[_]]
         }
       }
 
+      if (supportsCustomMultipartEncoding) {
+        "send a multipart message containing body parts with explicitly set encodings" in {
+
+          val utf8Char = "\u00f6"
+          val replacementChar = "\ufffd"
```
Comment (adamw) [inline core/src/test/scala/sttp/client3/testing/HttpTest.scala:417]: I think I would add a comment where this replacement char originates from (and maybe render both in comment, if they have some proper visualisations in utf-8?)

File: finagle-backend/src/main/scala/sttp/client3/finagle/FinagleBackend.scala (hunk 6 above)
Comment (dkarwacki) [reply finagle-backend/src/main/scala/sttp/client3/finagle/FinagleBackend.scala]: corrected

File: core/src/test/scala/sttp/client3/testing/HttpTest.scala (hunk 7 above)
Comment (dkarwacki) [reply core/src/test/scala/sttp/client3/testing/HttpTest.scala]: corrected

**Pull request merged**